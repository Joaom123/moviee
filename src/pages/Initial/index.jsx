import React, {Component} from 'react';
import {ModalMovie, Movies, Pagination, SearchForm, Snackbar} from "../../components";
import {apiService} from "../../services";

class Initial extends Component {
    constructor(props) {
        super(props);

        this.state = {
            searchValue: "",
            page: 1,
            movies: [],
            movie: {},
            totalResults: 0,
            toggleModal: false,
            errorMessage: "",
            isActive: false
        };

        this.handleChangeOfSearchValue = this.handleChangeOfSearchValue.bind(this);
        this.handleChangeOfPage = this.handleChangeOfPage.bind(this);
        this.onCardClick = this.onCardClick.bind(this);
    }

    onCloseModal = () => this.setState({toggleModal: false, movie: {}});

    doesGetMoviesListIfValidSearchValue = () => {
        if (this.isSearchValueValid())
            this.getMoviesListAndSetMovies();
    }

    doesGetMoviesListWithTimeout = () => {
        clearTimeout(this.timer);
        this.timer = setTimeout(
            () => this.doesGetMoviesListIfValidSearchValue(),
            800
        );
    }

    isSearchValueNotEmpty = () => this.state.searchValue !== "";

    isSearchValueLongerThan3Characters = () => this.state.searchValue.length >= 3;

    isSearchValueValid = () => this.isSearchValueNotEmpty() && this.isSearchValueLongerThan3Characters();

    getMoviesListAndSetMovies = () => {
        const {searchValue, page} = this.state;
        apiService
            .getMoviesList({searchValue, page})
            .then(response => {
                const {Response, Search, totalResults, Error} = response.data;

                if (Response === 'True') {
                    this.setState({
                        movies: Search,
                        totalResults
                    });
                } else {
                    this.setState({
                        errorMessage: Error,
                        isActive: true,
                        movies: [],
                        page: 1,
                        totalResults: 0
                    });
                }
            })
            .catch(error => {
                console.log(error);
            })
    }

    componentDidUpdate(prevProps, prevState, snapshot) {
        const {searchValue, page, isActive} = this.state;

        if(searchValue !== prevState.searchValue || page !== prevState.page)
            this.doesGetMoviesListWithTimeout();

        if(isActive === true && isActive !== prevState.isActive) {
            setTimeout(
                () => this.setState({isActive: false}),
                3000
            );

        }
    }

    handleChangeOfSearchValue(event) {
        let searchValue = event.target.value;

        if (searchValue === " ")
            return ;

        this.setState({searchValue});
    }

    handleChangeOfPage(e, page) {
        e.preventDefault();
        this.setState({page});
    }

    onCardClick(imdbId) {
        apiService
            .getMovieByImdbId({imdbId})
            .then(response => {
                this.setState({
                    movie: response.data,
                    toggleModal: true
                });
            });
    }

    render() {
        const {searchValue, movies, totalResults, page, movie, toggleModal, errorMessage, isActive} = this.state;

        return (
            <main className="container">
                <SearchForm
                    searchValue={searchValue}
                    handleChange={this.handleChangeOfSearchValue}
                />
                <Movies
                    movies={movies}
                    onCardClick={this.onCardClick}
                />
                <Pagination
                    totalResults={totalResults}
                    page={page}
                    handleChangeOfPage={this.handleChangeOfPage}
                />
                <ModalMovie
                    toggleModal={toggleModal}
                    onCloseModal={this.onCloseModal}
                    movie={movie}
                />
                <Snackbar
                    isActive={isActive}
                    message={errorMessage}
                />
            </main>
        );
    }
}

export default Initial;
