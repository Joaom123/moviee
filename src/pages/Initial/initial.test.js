import React from "react";
import { render, cleanup } from "@testing-library/react";
import Initial from "./index";

afterEach(cleanup);

it('snapshot ', () => {
    const {asFragment} = render(<Initial />);
    expect(asFragment(<Initial />)).toMatchSnapshot();
});