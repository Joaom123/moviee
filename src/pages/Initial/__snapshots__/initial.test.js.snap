// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot  1`] = `
<DocumentFragment>
  <main
    class="container"
  >
    <section
      class="row"
    >
      <form
        class="col s12"
      >
        <div
          class="row"
        >
          <div
            class="input-field col s12 m6 offset-m3"
          >
            <input
              class="validate"
              data-testid="searchMovieInput"
              id="movieName"
              name="movieName"
              type="text"
              value=""
            />
            <label
              for="movieName"
            >
              Title
            </label>
          </div>
        </div>
      </form>
    </section>
    <section
      class="moviesCards"
    />
    <section
      class="modal row modalMovie"
      id="modal"
      style="transform: perspective(300px) rotateY(0deg) rotateX(0deg); transition: transform 1s;"
      tabindex="0"
    />
    <span
      class="snackbar "
    />
  </main>
</DocumentFragment>
`;
